test/saithrift/tests/sail3.py

```diff
@@ -1214,3 +1214,114 @@ def runTest(self):
             self.client.sai_thrift_remove_router_interface(rif_id1)
             self.client.sai_thrift_remove_router_interface(rif_id2)
             self.client.sai_thrift_remove_virtual_router(vr_id)
+
+ @group('l3')
+class L3VlanNeighborMacUpdateTest(sai_base_test.ThriftInterfaceDataPlane):
+    def runTest(self):
+        """
+        For sai server for testing learning inside vlan and fwd packet to vlan through l3
+        Steps
+        1. Create VLAN
+        2. Create two VLAN members
+        3. Set port VLAN IDs
+        4. Create virtual router
+        5. Create router interface for VLAN and an extra port 3
+        6. Set SAI PORT LEARN MODE HW
+        7. Send packet from one port of the VLAN to the DUT to update the FDB entry
+        8. Create neighbor and route
+        9. Send L3 packet from port to the VLAN with the destination IP and verify that only the targeted port receives the packet and the MAC in the packet is updated 
+        10. clean up.
+        """
+
+        print
+        print "Sending packet port 1 -> switch (for learning purpuse)"
+        print "and then sending packet from port 3 -> port 1 (through the router)"
+        switch_init(self.client)
+        sai_thrift_clear_all_counters(self.client)
+        v4_enabled = 1
+        v6_enabled = 1
+        vlan_id = 10
+        port1 = port_list[0]
+        port2 = port_list[1]
+        port3 = port_list[2]
+        addr_family = SAI_IP_ADDR_FAMILY_IPV4
+        ip_addr1_subnet = '10.10.10.0'
+        ip_mask1 = '255.255.255.0'
+        mac_port1 = '00:0a:00:00:00:01'
+        ip_port1 = '10.10.10.2'
+        mac_port3 = '00:0b:00:00:00:01'
+        mac1 = ''
+        mac2 = ''
+        self.client.sai_thrift_create_vlan(vlan_id)
+        
+        vlan_member1 = sai_thrift_create_vlan_member(self.client, vlan_id, port1, SAI_VLAN_PORT_TAGGED)
+        vlan_member2 = sai_thrift_create_vlan_member(self.client, vlan_id, port2, SAI_VLAN_PORT_TAGGED)
+        attr_value1 = sai_thrift_attribute_value_t(u16=vlan_id)
+        attr1 = sai_thrift_attribute_t(id=SAI_PORT_ATTR_PORT_VLAN_ID, value=attr_value1)
+        self.client.sai_thrift_set_port_attribute(port1, attr1)
+        self.client.sai_thrift_set_port_attribute(port2, attr1)
+        
+        vr_id = sai_thrift_create_virtual_router(self.client, v4_enabled, v6_enabled)
+        rif_vlan_id = sai_thrift_create_router_interface(self.client, vr_id, 0, 0, vlan_id, v4_enabled, v6_enabled, mac1)
+        rif_port_id = sai_thrift_create_router_interface(self.client, vr_id, 1, port3, 0, v4_enabled, v6_enabled, mac2)
+
+        attr_value2 = sai_thrift_attribute_value_t(s32=SAI_PORT_LEARN_MODE_HW)
+        attr2 = sai_thrift_attribute_t(id=SAI_PORT_ATTR_FDB_LEARNING, value=attr_value2)
+        self.client.sai_thrift_set_port_attribute(port1, attr2)
+        self.client.sai_thrift_set_port_attribute(port2, attr2)
+        
+        sai_thrift_create_neighbor(self.client, addr_family, rif_vlan_id, ip_port1, mac_port1)
```

Comment:
I see a difference between the testcase summary and code implementation. Why the neighbor and route are created before sending packet to update the FDB entry? Could you provide some explanations?